### PR TITLE
Implement AWS key value store

### PR DESCRIPTION
**Pull request by ogghead**
Hi folks! I am creating this draft PR to solicit feedback on an initial AWS key value store implementation. I appreciate any and all discussions on this PR!

Some points for thought:
1. Implementation uses DynamoDB, though for large blob storage, S3 is preferable (DynamoDB can only store <=400KB size records), see https://github.com/fermyon/spin/issues/2606. DynamoDB is cheaper and faster for performing many rapid reads/writes of small amounts of data though, and is _roughly_ in the same niche as Azure CosmosDB
2. Auth currently requires generating AWS STS token credentials and passing them to the Spin app in a runtime config file. https://github.com/spinkube/skips/pull/9/files discusses better patterns to fetch credentials. Curious to hear thoughts on how this implementation can integrate better with that proposal!
3. The [Azure key-value implementation](https://github.com/fermyon/spin/blob/main/crates/key-value-azure/src/store.rs#L64) supports reading credentials from environment variables, however the [AWS Rust SDK does not offer a synchronous API to load config](https://docs.rs/aws-config/latest/aws_config/fn.load_from_env.html) and would require the [MakeKeyValueStore::make_store function](https://github.com/fermyon/spin/blob/main/crates/factor-key-value/src/runtime_config/spin.rs#L20) to be async for all implementations -- leading to a chain of async function coloring. It is possible to [manually fill the SdkConfig object](https://docs.rs/aws-config/latest/aws_config/struct.SdkConfig.html#method.builder) and I did this to pass STS tokens from a runtime config file, but it would be ideal to rely on the [SDK's defaults and many credential loading fallbacks if possible](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-authentication.html). Curious on thoughts for how to best handle env var credential loading

### CODE REVIEW

**Comment by itowlson** [comment]
Kia ora @ogghead and thanks for this.  We've got work going on in https://github.com/fermyon/spin/pull/2895 to implement some additional key-value interfaces, and I think it works better to land that one first, then have this PR include _all_ the AWS stuff.  This is partly down to what has the biggest compatibility implications combined with the present release timeline, but also will hopefully provide enough infrastructure to make extending this PR to the new interfaces easy!  I hope that's okay with you.

In the meantime, I'll try to have a look at your points for thought!

**Comment by itowlson** [comment]
> Implementation uses DynamoDB, though for large blob storage, S3 is preferable

I'd say this is fine for now.  We could call this a DynamoDB KV store, which would leave us flexibility to later to add a S3 KV backend if users had large object use cases - nothing here would preclude that as long as we think of it as an "AWS product X" store rather than an "AWS" store.

> Auth currently requires generating AWS STS token credentials

I don't think we are bound to offer a "tokens in the runtime config" option if that doesn't make sense or is painful to implement.  I'm not sure we can rely on the workload identity idea from that SKIP across all Spin runtime environments, but absolutely open to doing things differently as appropriate.

> it would be ideal to rely on the SDK's defaults and many credential loading fallbacks if possible

This is what we do in the SQS trigger.  There's no credential configuration, we just load the SDK and let figure out the credentials, whether from ambient EVs or whatever.  I'm told that's idiomatic enough, so I'd have no problem with doing the same thing here.  I'm sure someone will shout out if they do - but we could presumably retrofit additional configuration methods if need be - the Cosmos one certainly went through a few sets of extensions...

**Comment by ogghead** [comment]
> Kia ora @ogghead and thanks for this.  We've got work going on in https://github.com/fermyon/spin/pull/2895 to implement some additional key-value interfaces, and I think it works better to land that one first, then have this PR include _all_ the AWS stuff.  This is partly down to what has the biggest compatibility implications combined with the present release timeline, but also will hopefully provide enough infrastructure to make extending this PR to the new interfaces easy!  I hope that's okay with you.
> 
> 
> 
> In the meantime, I'll try to have a look at your points for thought!
> 
> 

Sounds good to me! I'm excited to see that work land. It makes sense to hold off on this for now, then rework after that is merged and support all WASI KV interfaces for AWS.

Thanks for taking a look!

**Comment by ogghead** [comment]
> > Implementation uses DynamoDB, though for large blob storage, S3 is preferable
> 
> 
> 
> I'd say this is fine for now.  We could call this a DynamoDB KV store, which would leave us flexibility to later to add a S3 KV backend if users had large object use cases - nothing here would preclude that as long as we think of it as an "AWS product X" store rather than an "AWS" store.

Great callout! The config specifies "Dynamo" as the KV store type, so should _hopefully_ be flexible to add other backends. I will keep this in mind when implementing the full WASI KV interface

> > Auth currently requires generating AWS STS token credentials
> 
> 
> 
> I don't think we are bound to offer a "tokens in the runtime config" option if that doesn't make sense or is painful to implement.  I'm not sure we can rely on the workload identity idea from that SKIP across all Spin runtime environments, but absolutely open to doing things differently as appropriate.
> 

The runtime config token setup is (from local testing) working -- but the "use the default SDK config loading" is proving challenging with the current interface constraints. Mainly as the default AWS config loader is implemented as an async function.

> > it would be ideal to rely on the SDK's defaults and many credential loading fallbacks if possible
> 
> 
> 
> This is what we do in the SQS trigger.  There's no credential configuration, we just load the SDK and let figure out the credentials, whether from ambient EVs or whatever.  I'm told that's idiomatic enough, so I'd have no problem with doing the same thing here.  I'm sure someone will shout out if they do - but we could presumably retrofit additional configuration methods if need be - the Cosmos one certainly went through a few sets of extensions...
> 
> 

Agreed, this is the pattern I followed in the Kinesis trigger as well. I would love to have this here too! The challenge is function coloring from AWS config default loader function -- using that async function here appeared to require a chain of refactoring across the general KV traits, but I must admit that my async Rust knowledge hit a wall when trying to reconcile the changes required for that.

**Comment by itowlson** [comment]
Ah, I misread your comment about using the default SDK config - sorry about that.

It seems like you could call Tokio's `block_on` to wrap the async call in a blocking wrapper.  It's not going to win any prizes for elegance but should, hopefully, get the job done.  The section "A synchronous interface to mini-redis" on the Tokio ["Bridging to sync code"](https://tokio.rs/tokio/topics/bridging) page seems like it might be close to what you want, although you may well have tried that already?  We do have some Giant Async Brains floating around who might be able to help if you share what you tried and what you ran into.

**Comment by ogghead** [comment]
> Ah, I misread your comment about using the default SDK config - sorry about that.
> 
> It seems like you could call Tokio's `block_on` to wrap the async call in a blocking wrapper. It's not going to win any prizes for elegance but should, hopefully, get the job done. The section "A synchronous interface to mini-redis" on the Tokio ["Bridging to sync code"](https://tokio.rs/tokio/topics/bridging) page seems like it might be close to what you want, although you may well have tried that already? We do have some Giant Async Brains floating around who might be able to help if you share what you tried and what you ran into.

Good callout -- I tried using `block_on` to get this working with
```
Handle::current().block_on(aws_config::load_defaults(BehaviorVersion::latest()))
```
as well as
```
let rt = tokio::runtime::Builder::new_current_thread()
                    .enable_all()
                    .build()?;
 rt.block_on(aws_config::load_defaults(BehaviorVersion::latest()))
```
While these _do_ compile, I see crashes immediately on Spin app startup with
```
Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.
```

Alternatively, when I went down the path of asyncifying everything required to await this function, I hit a wall at [store_from_toml_fn](https://github.com/ogghead/spin/blob/74e7705e421a820528b4513fd2b6bcc776bcfbee/crates/factor-key-value/src/runtime_config/spin.rs#L29) where closures are returned. [Async closure enhancements in Rust](https://rust-lang.github.io/rfcs/3668-async-closures.html) might be needed to make the closures returned there async, but this is where my knowledge of async Rust was lacking.

If any Giant Async Brains (or anyone) have ideas on the best path forward on this, much appreciated!

**Comment by itowlson** [comment]
All right.  I think I have a way around this for you.  "But," in the words of Deep Thought, "you're not going to like it."

So I asked the Giant Async Brains about blocking on `load_defaults` and they said "don't do that."  Instead the idea is to create a future for the Client and capture that in a lazy or once-cell instead.  Then await it each time you want to use it (which will be cheap after the first time, especially compared to the network activity that follows).

Here is what I did, which seems to work (but you may find a less awful way, this was just the first stab that didn't make the compiler mad at me):

* Add the `async-once-cell` crate
* Change `KeyValueAwsDynamo::client` to be a (deep breath) `async_once_cell::Lazy<Client, std::pin::Pin<Box<dyn std::future::Future<Output = Client> + Send>>>`
* In `KeyValueAwsDynamo::new`, shunt the existing code into an `async move`.  Then `Box::pin` the async move.  And capture that as a future.  Then put _that_ into a `Lazy::from_future`.  So it looks like:

```rust
        let client_fut: std::pin::Pin<Box<dyn std::future::Future<Output = Client> + Send>> = Box::pin(async move {
            let config = match auth_options {
                KeyValueAwsDynamoAuthOptions::RuntimeConfigValues(config) => /* as current */,
                KeyValueAwsDynamoAuthOptions::Environmental => {
                    aws_config::load_defaults(BehaviorVersion::latest()).await  // as before but uncommented
                }
            };
            Client::new(&config)
        });

        let client_cell = async_once_cell::Lazy::from_future(client_fut);

        Ok(Self { client: client_cell, table })
```

(Some of the naming here is poor, this was throwaway code.)

* Change `StoreManager::get` to `get_unpin().await` the Lazy:

```rust
    async fn get(&self, name: &str) -> Result<Arc<dyn Store>, Error> {
        Ok(Arc::new(AwsDynamoStore {
            _name: name.to_owned(),
            client: self.client.get_unpin().await.clone(),  // <-- this bit
            table: self.table.clone(),
        }))
    }
```

NOTE: this breaks `StoreManager::summary`.  You'll need to either make that async or capture the summary info as extra fields, but this should be routine.  (I hope.  I admit punting on this.)

Let me know if you need more info or want a proper diff.

**Comment by ogghead** [comment]
_Excellent_! This is exactly the Galaxy Async Brain thinking I was sorely lacking 😄

I will give this a go tonight, thanks for the tips!

**Comment by ogghead** [comment]
Can confirm this worked like a charm! Pushed the changes to reflect and I will keep an eye on the full WASI KV implementation PR. I am in your debt for your help on this @itowlson :)

One does not simply walk into `async_once_cell::Lazy<Client, std::pin::Pin<Box<dyn std::future::Future<Output = Client> + Send>>>` (or Mordor)

**Comment by itowlson** [comment]
I'm delighted to have helped!  Thanks once again for your effort, your patience, and your good humour throughout this...

...

...because you will need them when I call that debt in.  _ominous music and cheesy lightning FX in which the viewer can vaguely make out the looming shape of `wasi:blobstore`_

(also, and at the risk of bathos, please ignore MQTT CI failures - it's a known flake)

**Comment by ogghead** [comment]
I have implemented atomic and batch operations -- the logic is tentatively all in place. I will test the operations with a component using WASI-KV and then (barring any uncovered issues) mark this ready for review

**Review by itowlson — COMMENTED** [review]
Thanks for this!  I'm not very qualified to review the Dynamo stuff but @endocrimes has kindly volunteered to look at it.  In the meantime just a few comments and questions.

File: crates/key-value-aws/src/store.rs
Comment (ogghead) [reply crates/key-value-aws/src/store.rs:132]: Indeed, this appears to be dead -- further investigation shows it may also be dead in the Azure implementation (likely where I utilized my brilliant _ctrl+c, ctrl+v_ technique) -- I went ahead and removed this in both implementations

File: crates/key-value-aws/src/store.rs
Comment (ogghead) [reply crates/key-value-aws/src/store.rs:23]: That is fair -- the reasoning for `client` not being Arc is that client already wraps an Arc:
```
pub struct Client {
    handle: Arc<Handle>,
}
```

It should be low cost to clone the client without another Arc was my initial thought, but definitely open to thoughts!

File: crates/key-value-aws/src/store.rs
Comment (ogghead) [reply crates/key-value-aws/src/store.rs:185]: This is a good callout -- while I do not know of a specific operation meant to check whether an item with a specific key exists in DynamoDB, we can definitely avoid downloading the entire value by returning only a specific key (in this case, just the PK). I will update with this optimization, but I cannot speak for whether CosmosDB supports something similar

File: crates/key-value-aws/src/store.rs
Comment (ogghead) [inline crates/key-value-aws/src/store.rs:357]: Thinking aloud -- this can be brought closer to ensuring a unique lock: TransactWrite can return the VAL key while setting a lock key -- only under the condition that the lock key doesn't already exist. Combined with deleting the lock key on swap, this could guarantee that only one process can acquire "the lock" assuming all processes call `current` before `swap`. However, this does not prevent another process calling `swap` without first calling `current` and ignoring the lock. But, it may still be worth making this change to get as close as possible to a transaction -- curious to hear thoughts on the best approach for this!

File: crates/key-value-aws/src/store.rs
Comment (itowlson) [reply crates/key-value-aws/src/store.rs:23]: Yeah, it would have made sense to me if KeyValueAwsDynamo was Clone, but I didn't see a Clone implementation.  And although `Client` is clone, I don't think `async_once_cell::Lazy<Client, std:something::Terrifying<...>>` is, which now makes me wonder if this might be a hangover from a previous iteration?

File: crates/key-value-aws/src/store.rs
Comment (itowlson) [reply crates/key-value-aws/src/store.rs:185]: Sorry, the Cosmos mention was more a note to self to see what it does and put a note in the docs if it downloaded potentially large data.  Definitely not trying to put that on your plate.

File: crates/key-value-aws/src/store.rs
Comment (ogghead) [reply crates/key-value-aws/src/store.rs:23]: You are quite correct -- cloning these fields is only done in `StoreManager::get` and `StoreManager::summary` and the overall object is never cloned. But, I now realize that function itself returns an Arc -- so this may be unnecessary overhead.

In the `Store` implementation for `AwsDynamoStore` though, creating CAS handles through the `new_compare_and_swap` function made it seem ideal to use Arcs/low cost cloned objects for the client and table to allow creating many parallel CAS handles at low cost. That doesn't preclude containing owned data in the KeyValueAwsDynamo and creating an Arc for table in `StoreManager::get` though, would that be preferable?

Comment (itowlson) [reply crates/key-value-aws/src/store.rs:23]: I agree that the considerations for `AwsDynamoStore` are different - it was specifically this `KeyValueAwsDynamo` struct that was puzzling me.  But yeah if the expectation is that these fields will be repeatedly cloned then it makes sense to `Arc` them once here.  And with `client`, as you say, it's already cheap to await (except for the first time) and clone the result.  Thanks for the patient explanation - I'm happy now.

Comment (ogghead) [reply crates/key-value-aws/src/store.rs:23]: It is certainly possible that this is premature optimization and caching of the store is done in a table higher in the orchestration of key-value factors -- open to tweaking this! I appreciate the discussion :)

File: crates/key-value-aws/src/store.rs
Comment (ogghead) [reply crates/key-value-aws/src/store.rs:357]: Indeed it appears to be, so I will make this change which hopefully drastically simplify the calls in CAS

File: crates/key-value-aws/src/store.rs
Comment (endocrimes) [inline crates/key-value-aws/src/store.rs:475]: This would potentially be slightly easier to read with the SDK's paginator (https://docs.rs/aws-sdk-dynamodb/latest/aws_sdk_dynamodb/operation/scan/builders/struct.ScanFluentBuilder.html#method.into_paginator) - but I'm extremely not a Rustacean 😅  (and this otherwise is doing ~the same thing afaict)

**Review by endocrimes — COMMENTED** [review]
Dynamo uses eventually consistent reads by default - which is fine for applications written against Dynamo, but I believe our Spin KV contract is that you at least read-your-writes.

Dynamo charging 2x the cost for a Consistent Read makes this a compelling tradeoff regardless, but we may need to document that fact (and potentially make ConsistentReads an opt-in configuration?)

Otherwise this seems reasonably sound to me (modulo questions about CAS)

**Comment by ogghead** [comment]
> Dynamo uses eventually consistent reads by default - which is fine for applications written against Dynamo, but I believe our Spin KV contract is that you at least read-your-writes. Dynamo charging 2x the cost for a Consistent Read makes this a compelling tradeoff regardless, but we may need to document that fact (and potentially make ConsistentReads an opt-in configuration?)
> 
> Otherwise this seems reasonably sound to me (modulo questions about CAS) - thanks for the PR!

Good point! I was mulling over whether to add a configuration to specify strongly consistent reads, and this comment makes it clear that would be good to have (and potentially default to strongly consistent to maintain consistency of behavior with other implementations?) It should be quick to add, will push that up either this morning or tonight

File: crates/key-value-aws/src/store.rs
Comment (ogghead) [reply crates/key-value-aws/src/store.rs:475]: Good point -- I was unaware of this utility! I will look into replacing custom logic with the paginator.

**Comment by ogghead** [comment]
Ok! I may have gotten a bit pedantic with the states for Cas but this is hopefully in line with what you were imagining @endocrimes and cleans up. That said, I uncovered some odd behavior while testing and published an "int test Spin app" [repo here](https://github.com/ogghead/TestDynamo) for reference
1. For AWS, I observed panics [here](https://github.com/ogghead/TestDynamo/blob/main/src/lib.rs#L104) 
```
Caused by:
    0: error while executing at wasm backtrace:
           0: 0x4812a - wit-component:shim!indirect-wasi:keyvalue/atomics@0.2.0-draft2-swap
           1: 0xf034 - test_dynamo.wasm!spin_sdk::wit::wasi::keyvalue::atomics::swap::hfb1256b17b155b22
           2: 0x64fc - test_dynamo.wasm!test_dynamo::handle_test_dynamo::hb0dc3a6709598c91
           3: 0x77ba - test_dynamo.wasm!spin_executor::run::h6238cdfeaf27259c
           4: 0x954c - test_dynamo.wasm!wasi:http/incoming-handler@0.2.0#handle
    1: CasError::CasFailed(Resource { rep: 4, state: "own (not in table)" })
```
2. For sanity, I tested with sqlite to see the "reference behavior" but there I hit [panics here](https://github.com/ogghead/TestDynamo/blob/main/src/lib.rs#L47)
```
Caused by:
    0: error while executing at wasm backtrace:
           0: 0x1bd0e - test_dynamo.wasm!__rust_start_panic
           1: 0x1bba7 - test_dynamo.wasm!rust_panic
           2: 0x1bad5 - test_dynamo.wasm!std::panicking::rust_panic_with_hook::h6e665b71c8f50b27
           3: 0x1ac0f - test_dynamo.wasm!std::panicking::begin_panic_handler::{{closure}}::hb63ceb92de73cef0
           4: 0x1ab75 - test_dynamo.wasm!std::sys::backtrace::__rust_end_short_backtrace::hf18948010daec5d9
           5: 0x1b3b1 - test_dynamo.wasm!rust_begin_unwind
           6: 0x299bf - test_dynamo.wasm!core::panicking::panic_fmt::h7916dc01f99baff2
           7: 0x2b1cd - test_dynamo.wasm!core::panicking::assert_failed_inner::h636831bd2590e950
           8: 0x3538 - test_dynamo.wasm!core::panicking::assert_failed::h4b4824a9d77e94de
           9: 0x51e8 - test_dynamo.wasm!test_dynamo::handle_test_dynamo::hb0dc3a6709598c91
          10: 0x77ba - test_dynamo.wasm!spin_executor::run::h6238cdfeaf27259c
          11: 0x954c - test_dynamo.wasm!wasi:http/incoming-handler@0.2.0#handle
    1: wasm trap: wasm `unreachable` instruction executed
```

It is certainly possible that these issues are caused by my own environment but the AWS one appears to potentially be rooted in higher level orchestration logic [around atomic retries here](https://github.com/ogghead/spin/blob/main/crates/factor-key-value/src/host.rs#L388)

File: crates/key-value-aws/src/store.rs
Comment (rylev) [inline crates/key-value-aws/src/store.rs:213]: Why do we need to check for key existence? Does `client.delete_item` fail if the item doesn't exist? If so, we still have a race condition between the call to `exists` and `delete_item` where the item might be deleted and the error could occur.

File: crates/key-value-aws/src/store.rs
Comment (rylev) [inline crates/key-value-aws/src/store.rs:252]: Nit: `with_capacity` does allocation which we can avoid in the empty key case by moving the initialization down.

```suggestion
        if keys.is_empty() {
            return Ok(Vec::new());
        }
        
         let mut results = Vec::with_capacity(keys.len());
```

Comment (ogghead) [reply crates/key-value-aws/src/store.rs:252]: As a side note, I see [potentially unnecessary vec allocations at that level](https://github.com/ogghead/spin/blob/main/crates/factor-key-value/src/host.rs#L287)

File: crates/key-value-aws/src/store.rs
Comment (ogghead) [reply crates/key-value-aws/src/store.rs:213]: Good point! It appears that the delete operation only fails if a condition is set. I will remove this check as the API does not throw errors when running delete on nonexistent items

**Comment by ogghead** [comment]
> > cc @devigned for the possible issue in the SQLite (or host?) implementation ([#2883 (comment)](https://github.com/fermyon/spin/pull/2883#issuecomment-2472390361))
> 
> I'm at KubeCon right now, but I will try to give it a look when I have a break. At first glance, it seems like the CAS resource is not registered in the CAS resource table. The test code looked correct to me, so this is likely a bug in the CAS implementation.

I appreciate the quick responses! On lunch so pushed up a few fixes for latest comments. I did some additional testing and the issue on sqlite get_many does appear to be inconsistent and I can reproduce with AWS too -- I have a hunch that [caching of deleted values](https://github.com/ogghead/spin/blob/509d8d176cda57c9f7b36db5738fd3a0fc41d30f/crates/factor-key-value/src/util.rs#L286) could be involved as adding sleeps in between delete_many and get_many operations allows consistent passes of that test for sqlite.

I did end up adding one fallback in the sqlite implementation in `swap` for null old_value as I was seeing an [edge case crash here](https://github.com/ogghead/spin/blob/509d8d176cda57c9f7b36db5738fd3a0fc41d30f/crates/key-value-spin/src/store.rs#L320), please let me know if the modified behavior isn't desirable! With this change in place, I can confirm that both sqlite and AWS crash at the same place -- though I have yet to trace why the error including the new CAS isn't being returned user-side and instead logs/crashes somewhere

File: crates/key-value-aws/src/store.rs
Comment (ogghead) [reply crates/key-value-aws/src/store.rs]: I made the changes at the higher level (and removed vec copying there) -- hopefully this makes sense, let me know if it's preferable to leave this empty list check at the AWS level

File: crates/key-value-aws/src/store.rs
Comment (ogghead) [reply crates/key-value-aws/src/store.rs:469]: Updated using the logic you suggested combined with what I had previously (using version key) -- thanks for the feedback!

**Review by itowlson — APPROVED** [review]
This looks good to me - I'll wait for Danielle to weigh in though!  Thank you for this epic effort!

**Review by endocrimes — APPROVED** [review]
This looks great - thanks!

(and oh wow, that ended up being a real can of 🪱 - double thanks for chasing those down!)

**Pull request merged**
